src/firmware/app/control/trajectory_planner.c

```diff
@@ -0,0 +1,356 @@
+        traj_elements[i].position = shared_polynomial2d_getValueOrder3(path, t);
+    }
+}
+
+void app_trajectory_planner_getMaxAllowableSpeedProfile(
+    float max_allowable_speed_profile[TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS],
+    Polynomial2dOrder3_t path, unsigned int num_elements,
+    ArcLengthParametrization_t arc_length_parameterization, float arc_segment_length,
+    const float max_allowable_acceleration)
+{
+    Polynomial2dOrder2_t first_deriv = shared_polynomial2d_differentiateOrder3(path);
+    Polynomial2dOrder1_t second_deriv =
+        shared_polynomial2d_differentiateOrder2(first_deriv);
+
+    for (unsigned int i = 0; i < num_elements; i++)
+    {
+        // Get the 't' value corresponding to the current arc length (to be used for
+        // further computing)
+        float t = shared_polynomial2d_getTValueAtArcLengthOrder3(
+            path, i * arc_segment_length, arc_length_parameterization);
+
+        // Create the polynomial representing path curvature
+        //                                              1
+        //                              ---------------------------------
+        //                                     abs(x'y'' - y'x'')
+        //        radius of curvature =      ----------------------
+        //                                     (x'^2 + y'^2)^(3/2)
+        //
+        const float numerator =
+            fabs(shared_polynomial1d_getValueOrder2(first_deriv.x, t) *
+                     shared_polynomial1d_getValueOrder1(second_deriv.y, t) -
+                 shared_polynomial1d_getValueOrder2(first_deriv.y, t) *
+                     shared_polynomial1d_getValueOrder1(second_deriv.x, t));
+        const float denominator =
+            pow(pow(shared_polynomial1d_getValueOrder2(first_deriv.x, t), 2) +
+                    pow(shared_polynomial1d_getValueOrder2(first_deriv.y, t), 2),
+                3.0 / 2.0);
+
+        const float radius_of_curvature = 1 / (numerator / denominator);
+
+        max_allowable_speed_profile[i] =
+            sqrt(max_allowable_acceleration * radius_of_curvature);
+    }
+}
+
+void app_trajectory_planner_generateForwardsContinuousVelocityProfile(
+    unsigned int num_segments,
+    float velocity_profile[TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS],
+    float max_allowable_speed_profile[TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS],
+    const float arc_segment_length, const float max_allowable_acceleration,
+    const float max_allowable_speed)
+{
+    for (unsigned int i = 1; i < num_segments; i++)
+    {
+        // Vf = sqrt( Vi^2 + 2*constant_segment_length*max_acceleration)
+        const float temp_vel =
+            (float)sqrt(pow(velocity_profile[i - 1], 2) +
+                        2 * arc_segment_length * max_allowable_acceleration);
+
+        // If the new speed is greater than the max allowable, reduce it to the max
+        if (temp_vel > max_allowable_speed_profile[i])
```

Comment:
Just going to point this out because it took me a sec to figure out, and may or may not trip up future people with less context.

The only reason this if-block is correct is because we can safely assume that all values in the `max_allowable_speed_profile` are `< max_allowable_speed`. Because we are using `else-if`, if `max_allowable_speed_profile < temp_vel < max_allowable_speed` and `max_allowable_speed_profile` was larger than the physical limitation of the robot, this would be a bug.

I'd suggest either mentioning this assumption just to be extra-explicit. Alternatively, I think it may be more clear to others to write

```
velocity_profile[i] = (float)sqrt(pow(velocity_profile[i - 1], 2) + 2 * arc_segment_length * max_allowable_acceleration);

if(velocity_profile[i] > max_allowable_speed_profile[i]) {
	velocity_profile[i] = max_allowable_speed_profile[i];
}
if(velocity_profile[i] > max_allowable_speed) {
	velocity_profile[i] = max_allowable_speed;
}
```
Or something like this to avoid the `else-if`, because then these assumptions don't really matter. It's more obvious we are applying limitations in either case. It's a nitpicky change based on what I think might trip up future people looking at this code, so definitely would like opinions on this.

@garethellis0 @EvanMorcom @jonathanlew @phelanm20

Reply:
max_allowable_speed_profile defines the maximum physically possible speed at each point that the robot can remain on the profile without slipping off.

It is the upper limit possible, and clamps the velocity to be lower than it even if max_allowable_speed is greater than it.

Reply:
maybe clarify the comment a bit more? It looked fine to me

Reply:
Right, but if max_allowable_speed` was less than `max_allowable_speed_profile` then it would be the limiting factor.

I realized I messed up my above example.

If `max_allowable_speed_profile` and `max_allowable_speed` are **BOTH** `< temp_vel`, then we are not guaranteed to clamp `temp_vel` to the smaller one because of the if-else statement.

Eg. If
```
temp_vel = 5;
max_allowable_speed_profile = 3;
max_allowable_speed = 2;
```
then after the if-block `velocity_profile[i] = max_allowable_speed_profile = 3`, even though we specified `max_allowable_speed = 2` and would expect that to be the velocity value.

So there actually is a bug here because of the `if-else`. We really want to limit `temp_vel` to `min(max_allowable_speed, max_allowable_speed_profile)`

I think what I suggested above is more correct (because C unfortunately doesn't have a `min` function for floats AFAIK):
```
velocity_profile[i] = (float)sqrt(pow(velocity_profile[i - 1], 2) + 2 * arc_segment_length * max_allowable_acceleration);

if(velocity_profile[i] > max_allowable_speed_profile[i]) {
	velocity_profile[i] = max_allowable_speed_profile[i];
}
if(velocity_profile[i] > max_allowable_speed) {
	velocity_profile[i] = max_allowable_speed;
}
```

Reply:
Fixed :+1: